### ISSUE TITLE
Cannot type chat when modal window is open (eg: formspec)

### ISSUE THREAD

**Issue by WSDguy2014** [minetest/minetest#3928]
In multiplayer, i can't type a chat, while sleeping, im trying to open console or opening chat, while sleeping, i can't. Take a look!
![minetest bug](https://cloud.githubusercontent.com/assets/13905392/14104985/726cd616-f5b1-11e5-8cbd-5b99ad14f0ea.png)

In Minecraft, you can able to type a chat, while sleeping.
![you can type chat while sleeping in minecraft](https://cloud.githubusercontent.com/assets/13905392/14105008/96a5754c-f5b1-11e5-98d2-0f13161e3579.png)
It should be have text box, to typing it on chat.

**Comment by kilbith** [comment]
You can't open the chat console while you have a formspec opened. That's impossible currently.

**Comment by WSDguy2014** [comment]
It should be update gui, while sleeping, adding text box and "Proceed" button, to say it.

**Comment by tenplus1** [comment]
Instead of formspec why not show in chat WHO is sleeping as well as player information like (2/10) 20% sleeping...  then player has to click to get out of bed or wait until over half the players are sleeping then daytime and ejected from bed ?

**Comment by paramat** [comment]
This is because you're sleeping :) Nevermind Minecraft.

**Comment by WSDguy2014** [comment]
@paramat, I found way, its possible to sleep at occupied bed.

**Comment by C1ffisme** [comment]
Don't you see the user that is spamming 'Hi'?

To be fair, I can see a use for this, if you need to say 'Go to bed!' to another person on your LAN server, but it's not very realistic, and usually 10 people on a server cannot agree to all sleep in their beds rather than kill mobs or just keep building.

**Comment by WSDguy2014** [comment]
@C1ffisme, i know. but, i forgot...

**Comment by paramat** [comment]
Good point about saying 'come to bed' to others.

**Comment by C1ffisme** [comment]
We could print a chat message if someone gets in a bed, something like:

```(Player) is in bed, (number) more people must get in bed to skip the night!```

**Comment by C1ffisme** [comment]
More than a year later, it seems like the demand for this is purely to match Minecraft, and it's not really useful for anything other than, well, chatting for the few seconds you are in bed.

However, I do think that allowing a chat interface in formspecs might be useful. Especially with complicated formspecs, it might be useful for a player to chat with his friends on how to use the GUI.

**Comment by orbea** [comment]
This is not to only match minecraft. Laying in bed for extended periods of time and being able to chat to other players is a much desired use case in our map. It helps a lot with coordinating other players into a bed to skip night.

**Comment by C1ffisme** [comment]
@orbea The problem with that is that you can simply type up a message out of bed, then immediately get in. No need for this chat menu.

**Comment by Digit** [comment]
> it seems like the demand for this is purely to match Minecraft, 

nonsense irrelevance.  

> and it's not really useful for anything other than, well, chatting for the few seconds you are in bed.

and how useful is bed then?  no one wants to go into bed when you can do _nothing_ while in there, except get out.  ability to co-ordinate from bed is much more sensible.  makes beds useful, welcoming, and prevents them being an unpleasant inconvenience being denied interactive connection to the world.

hopping in and out of bed to engage with others not yet in bed, to check if they are going to bed... it's ridiculous.  really is the biggest annoyance in minetest.  seems no good reason to deny players access to chat in that situation, where it might actually be the only time they need to chat and co-ordinate with others.

similarly while looking in the guide, or inventory, chests, etc, where it might be useful to converse about things.  cant keep it open while typing about it, cant mouse over n click about to check, mid sentence, without interuption having to close one to access the other.    chat could easily still drop down.  idk the logic why its suppressed.

seems maddening that at these times you'd most want to chat, you'd denied the ability to chat.    whyyyyyyyyyyyyy.    just to try to be different to minecraft?   idgaf about minecraft.  lets just have sensible consistent freedom to chat, esp when we most want to.   XD   maddening.  maddening.  maddening.

**Comment by orbea** [comment]
@C1ffisme And then get back out again to say something else, then back in, back out, back in, back out..... I hope you can see the frustration in this? :)

**Comment by C1ffisme** [comment]
> and how useful is bed then? no one wants to go into bed when you can do nothing while in there, except get out. ability to co-ordinate from bed is much more sensible. makes beds useful, welcoming, and prevents them being an unpleasant inconvenience being denied interactive connection to the world.

The point of the bed isn't so that you can do stuff while you are in it. It's to change a boolean that says whether or not a player is sleeping. If we're talking about singleplayer, this is 6 seconds that you have to wait. If you can't wait 6 seconds, I'm afraid you'll just need to blame your attention span.

> hopping in and out of bed to engage with others not yet in bed, to check if they are going to bed... it's ridiculous. really is the biggest annoyance in minetest. seems no good reason to deny players access to chat in that situation, where it might actually be the only time they need to chat and co-ordinate with others.

I have an idea... Just wait to get in bed until they say "I have made it to bed". There's no need for you to jump in and out of bed repeatedly.

> similarly while looking in the guide, or inventory, chests, etc, where it might be useful to converse about things. cant keep it open while typing about it, cant mouse over n click about to check, mid sentence, without interuption having to close one to access the other. chat could easily still drop down. idk the logic why its suppressed.

Which is why I suggest that being able to use chat inside a formspec can be useful. But I won't be needing chat for the 30 seconds that I'm interacting with a chest, furnace or inventory.

> seems maddening that at these times you'd most want to chat, you'd denied the ability to chat. whyyyyyyyyyyyyy. just to try to be different to minecraft? idgaf about minecraft. lets just have sensible consistent freedom to chat, esp when we most want to. XD maddening. maddening. maddening. chat in that situation, where it might actually be the only time they need to chat and co-ordinate with others.

I'm trying as hard as I can not to talk more about how you've typed rather than what you've typed.


I'm not totally object to having chat in the bed or any other inventory, but I think it's better that we add API functionality in the first place before adding this specific usecase.

**Comment by orbea** [comment]
As for wanting to copy minecraft, the biggest reason that this argument is invalid is that some of us still remember having this exact same problem in early releases of minecraft and wanting it to work back then too. If I recall it was soon after.

**Comment by orbea** [comment]
This use case is primarily for multiplayer, singleplayer is not relevant. :)  Coordinating from out of bed can work, but its highly inefficient, unreliable and annoying. The need to use it in a furnace, chest or the inventory is not quite as much as in a bed, but I think it will still will be a widely used feature if implemented.

And of course I agree that this should be implemented correctly, but I think that it stands that it should be implemented.

**Comment by Digit** [comment]
> a boolean that says whether or not a player is sleeping.

peculiar confusion.  in bed.  sleeping.  to me these seem like two distinct states.   sleep, is effectively an unconscious state without the passage of time.   one can lay in bed, awake, fully aware of the passage of time.  if night passes by, then one can say sleep was achieved.  if night drags on, doesnt get skipped, one cannot really say you slept.   you might argue the ingame character is sleeping, even though the real world player still has to suffer that passage of time, twiddling their thumbs... and oh wouldnt that be nice to be able to chat while in there waiting on others, to help pass the time, to engage with the latest shenanigans, coordinations, news etc... and once everyone gets to bed, night skips.  no need to interupt those conversations with checking if everyone is standing beside a bed ready to go to bed.   what a unecessary fuss and delay.  so much so in practice it just doesnt get used.

> single player

so?  why does that feature in your reply?  who cares about 6 seconds in single player... who cares about chat in single player?

> I have an idea... Just wait to get in bed until they say "I have made it to bed". There's no need for you to jump in and out of bed repeatedly.

your idea is just presenting the same problem in another way.

**Comment by DS-Minetest** [comment]
A simple issue causes much text again. \:P

(Another problem is with beds that while you type your message, the others might go to bed without you.)

How difficult would it be to allow opening chat or console while a formspec is shown?

**Comment by C1ffisme** [comment]
> peculiar confusion. in bed. sleeping. to me these seem like two distinct states. sleep, is effectively an unconscious state without the passage of time. one can lay in bed, awake, fully aware of the passage of time. if night passes by, then one can say sleep was achieved. if night drags on, doesnt get skipped, one cannot really say you slept. you might argue the ingame character is sleeping, even though the real world player still has to suffer that passage of time, twiddling their thumbs... and oh wouldnt that be nice to be able to chat while in there waiting on others, to help pass the time, to engage with the latest shenanigans, coordinations, news etc... and once everyone gets to bed, night skips. no need to interupt those conversations with checking if everyone is standing beside a bed ready to go to bed. what a unecessary fuss and delay. so much so in practice it just doesnt get used.

If you want me to read something, can you at least condense it into a few sentences that tell me exactly and precisely what you want me to hear, rather than giving me a bunch of words crammed into a paragraph that is longer than it needs to be?

(Add to that the fact that it's extra hard to find where one sentence ends and another begins, because you didn't capitalize the first letter of any of your sentences.)

As I have already said, _I am okay with this feature,_ but I _do not think it is high priority._ What is higher priority is adding functionality to allow chat in a formspec at all.

**Comment by orbea** [comment]
I read it perfectly fine, can we please discuss the content of the argument instead? :)

**Comment by ExeterDad** [comment]
Chatting while in bed definitely has a valid use.
I've yet to see a server were a bed is used to make it daytime again as coded. 
I believe the reason is simply because nobody can count on the majority to use the bed because a player may have to wait indefinitely for enough players to join in sleeping. So instead, beds are a decoration, or used for a place to spawn when killed. Nothing more.
Having access to a method of chatting to convince other players to go to bed to make it daylight would be very helpful.  **"Hey, we only need two more players to sleep and it will become daytime"** Might actually convince others to stop what they are doing and join.

**Comment by DS-Minetest** [comment]
>As I have already said, _I am okay with this feature,_ but I _do not think it is high priority_. What is higher priority is adding functionality to allow chat in a formspec at all.

What feature do you mean? Afaik this issue is only about adding the functionality to allow chat in a formspec at all.
>Cannot type chat when modal window is open (eg: formspec)

**Comment by ExeterDad** [comment]
`Cannot type chat when modal window is open (eg: formspec)`
It doesn't have to be "official chat".
A text input with a "Send" or "Chat" button will fire off a message to appear in chat without the formspec even closing.

**Comment by orbea** [comment]
> I've yet to see a server were a bed is used to make it daytime again as coded. 
> I believe the reason is simply because nobody can count on the majority to use the bed because a player may have to wait indefinitely for enough players to join in sleeping. So instead, beds are a decoration, or used for a place to spawn when killed. Nothing more.

In small private server which only has a few players it can work this way. :)

**Comment by paramat** [comment]
Duplicate of #3353
